network/alsp/manager/manager.go

```diff
@@ -0,0 +1,160 @@
+		m.logger.Warn().Msg("penalty mechanism of alsp is disabled")
+		return m
+	}
+
+	m.cache = internal.NewSpamRecordCache(cfg.SpamRecordsCacheSize, cfg.Logger, cfg.CacheMetrics, model.SpamRecordFactory())
+
+	for _, opt := range opts {
+		opt(m)
+	}
+
+	return m
+}
+
+// HandleMisbehaviorReport is called upon a new misbehavior is reported.
+// The current version is at the minimum viable product stage and only logs the reports.
+// The implementation of this function should be thread-safe and non-blocking.
+// TODO: the mature version should be able to handle the reports and take actions accordingly, i.e., penalize the misbehaving node
+// and report the node to be disallow-listed if the overall penalty of the misbehaving node drops below the disallow-listing threshold.
+func (m *MisbehaviorReportManager) HandleMisbehaviorReport(channel channels.Channel, report network.MisbehaviorReport) {
+	lg := m.logger.With().
+		Str("channel", channel.String()).
+		Hex("misbehaving_id", logging.ID(report.OriginId())).
+		Str("reason", report.Reason().String()).
+		Float64("penalty", report.Penalty()).Logger()
+	m.metrics.OnMisbehaviorReported(channel.String(), report.Reason().String())
+
+	if m.disablePenalty {
+		// when penalty mechanism disabled, the misbehavior is logged and metrics are updated,
+		// but no further actions are taken.
+		lg.Trace().Msg("discarding misbehavior report because ALSP module is disabled")
+		return
+	}
+
+	applyPenalty := func() (float64, error) {
+		return m.cache.Adjust(report.OriginId(), func(record model.ProtocolSpamRecord) (model.ProtocolSpamRecord, error) {
+			if report.Penalty() > 0 {
+				// this should never happen, unless there is a bug in the misbehavior report handling logic.
+				// we should crash the node in this case to prevent further misbehavior reports from being lost and fix the bug.
+				// TODO: refactor to throwing error to the irrecoverable context.
+				lg.Fatal().Float64("penalty", report.Penalty()).Msg(FatalMsgNegativePositivePenalty)
+			}
+			record.Penalty += report.Penalty() // penalty value is negative. We add it to the current penalty.
+			return record, nil
+		})
+	}
+
+	init := func() {
+		initialized := m.cache.Init(report.OriginId())
+		lg.Trace().Bool("initialized", initialized).Msg("initialized spam record")
+	}
+
+	// we first try to apply the penalty to the spam record, if it does not exist, cache returns ErrSpamRecordNotFound.
+	// in this case, we initialize the spam record and try to apply the penalty again. We use an optimistic update by
+	// first assuming that the spam record exists and then initializing it if it does not exist. In this way, we avoid
+	// acquiring the lock twice per misbehavior report, reducing the contention on the lock and improving the performance.
+	updatedPenalty, err := internal.TryWithRecoveryIfHitError(internal.ErrSpamRecordNotFound, applyPenalty, init)
```

Comment:
We can remove the func altogether and utilize the retry lib already in place we can accomplish the same thing without having to maintain this separate semi generic func. Something like this in place should suffice. 

```
	ctx := context.Background()
	backoff := retry.NewConstant(3*time.Second)
	retry.WithMaxRetries(2, backoff)
	err := retry.Do(ctx, backoff, func(context.Context) error {
		err := applyPenalty()
		switch {
		case err == nil:
			return nil
		case errors.Is(err, internal.ErrSpamRecordNotFound):
                         init() // call recovery func
			return retry.RetryableError(err)
		default:
			return err
		}
	})
	if err != nil {
		return err
	}
```

Reply:
https://github.com/onflow/flow-go/pull/4319/commits/217cff8092f1e274e48edd94325c490e57744cd1